src/background.js

```diff
@@ -25,6 +25,49 @@ async function initApi() {
   return api !== null;
 }
 
+/* Dynamic badge */
+function setStarredBadge(tabId) {
+  browser.browserAction.setBadgeText({ text: "★", tabId: tabId });
```

Comment:
This API doesn't work in Chrome, where Manifest V3 is used:
```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'getBadgeText')
```

I would suggest to move the logic for setting the badge into `browser.js` and provide separate implementations for Chrome and Firefox using either `browser.action` (Chrome) or `browser.browserAction` (Firefox).

Reply:
@sissbruecker Regarding this, I wonder why we have a separate branch for chrome with manifest v3, and on master using v2, as this is related to which manifest version you are using. As of [firefox version 109](https://extensionworkshop.com/documentation/develop/manifest-v3-migration-guide/), it supports manifest version v3. Version 109 was released in January 2023.

Reply:
See discussion here: https://github.com/sissbruecker/linkding-extension/pull/47#discussion_r1278341611

IIRC it changes the permission model to be more granular (per domain), and Firefox shows a blue dot below the extension icon for each web page that the user has not given the extension permission for. To make things more confusing, the permission is automatically granted after clicking the extension icon, and the dot just disappears, leaving the user confused what the dot was even about. Then the dot shows up again for every other web page.

There is a way for the extension to request additional permissions for all domains after installation, but that just adds more friction and isn't really better than the v2 way of doing things. It also wasn't clear if requesting permissions for all domains was an intentional use of the API or not. 

Long story short, the way v3 works in Firefox makes the UX worse, at which point I decided to maintain the different implementations between Chrome and FF.

Reply:
Regarding the chrome branch, it only has a modified `manifest.json`, the remaining code in the main branch is made compatible with Manifest v3 by extracting browser-specific API into `browser.js`.

Reply:
Thank you, it sounds like for the time being keeping mv2 for Firefox is the way to go.

I found this forum post about the permissions api which you should use to listen if permissions has changed. https://discourse.mozilla.org/t/request-permissions-at-install-time-manifest-v3/121543/7

I assume that at some point mv2 will be deprecated so worth keeping an eye out. Though Mozilla hasn't yet set a timeline so a deprecation seems to be quite far into the future.

Reply:
Extracted APIs for interacting with the badge into `browser.js`